java-frontend/src/main/java/org/sonar/java/se/checks/CustomUnclosedResourcesCheck.java

```diff
@@ -0,0 +1,289 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.MethodMatcherFactory;
+import org.sonar.java.se.CheckerContext;
+import org.sonar.java.se.ProgramState;
+import org.sonar.java.se.constraint.ConstraintManager;
+import org.sonar.java.se.constraint.ObjectConstraint;
+import org.sonar.java.se.symbolicvalues.SymbolicValue;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.NoSqale;
+import org.sonar.squidbridge.annotations.RuleTemplate;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+@Rule(
+  key = "S3546",
+  name = "Resources as defined by user should be closed",
+  priority = Priority.BLOCKER,
+  tags = {"bug", "cert", "cwe", "denial-of-service", "leak", "security"})
+@RuleTemplate
+@NoSqale
+public class CustomUnclosedResourcesCheck extends SECheck {
+
+  enum Status {
+    OPENED, CLOSED
+  }
+
+  @RuleProperty(
+    key = "constructor",
+    description = "the fully-qualified name of a constructor that creates an open resource."
+      + " An optional signature may be specified after the class name."
+      + " E.G. \"org.assoc.res.MyResource\" or \"org.assoc.res.MySpecialResource(java.lang.String, int)\"")
+  public String constructor = "";
+
+  @RuleProperty(
+    key = "factoryMethod",
+    description = "the fully-qualified name of a factory method that returns an open resource, with or without a parameter list."
+      + " E.G. \"org.assoc.res.ResourceFactory#create\" or \"org.assoc.res.SpecialResourceFactory #create(java.lang.String, int)\"")
+  public String factoryMethod = "";
+
+  @RuleProperty(
+    key = "openingMethod",
+    description = "the fully-qualified name of a method that opens an existing resource, with or without a parameter list."
+      + " E.G. \"org.assoc.res.ResourceFactory#create\" or \"org.assoc.res.SpecialResourceFactory #create(java.lang.String, int)\"")
+  public String openingMethod = "";
+
+  @RuleProperty(
+    key = "closingMethod",
+    description = "the fully-qualified name of the method which closes the open resource, with or without a parameter list."
+      + " E.G. \"org.assoc.res.MyResource#closeMe\" or \"org.assoc.res.MySpecialResource#closeMe(java.lang.String, int)\"")
+  public String closingMethod = "";
+
+  private List<MethodMatcher> classConstructor;
```

Comment:
Instead of using Lists of ```MethodMatcher```, please use ```MethodInvocationMatcherCollection``` (which should be renamed as ```MethodMatcherCollection```). 

This would allow you to use method ```anyMatch(Tree tree)``` from the MethodMatcherCollection instead of having to redefine methods at lines [L178](https://github.com/SonarSource/sonar-java/pull/710/files#diff-380c84c700fc90df37072a5b3c8b9c76R178) and [L187](https://github.com/SonarSource/sonar-java/pull/710/files#diff-380c84c700fc90df37072a5b3c8b9c76R187)